source/rpt/rpt_fem_reconstruction.cc

```diff
@@ -0,0 +1,1080 @@
+
+      // 4th constraint on the location of the particle in reference coordinates
+      last_constraint_reference_location = 1 - reference_location[0] -
+                                           reference_location[1] -
+                                           reference_location[2];
+
+      // Evaluate the error of the reference position (Is it outside the
+      // reference tetrahedron ?)
+      norm_error_coordinates =
+        calculate_reference_location_error(reference_location,
+                                           last_constraint_reference_location);
+
+      // Extrapolation limit
+      if (norm_error_coordinates < tol_reference_location)
+        {
+          // Calculate cost with the selected cost function
+          calculated_cost = calculate_cost(adjacent_cell,
+                                           reference_location,
+                                           last_constraint_reference_location,
+                                           experimental_count);
+          // Check if the new cost is lower than the previously stored
+          // real_location's
+          if (calculated_cost < max_cost_function)
+            {
+              position_found         = true;
+              max_cost_function      = calculated_cost;
+              previous_position_cell = adjacent_cell;
+
+              // Evaluate the real location of the particle
+              for (unsigned int v = 0; v < adjacent_cell->n_vertices(); ++v)
+                {
+                  if (v == 0)
+                    real_location = adjacent_cell->vertex(v);
+                  else
+                    real_location +=
+                      reference_location[v - 1] *
+                      (adjacent_cell->vertex(v) - adjacent_cell->vertex(0));
+                }
+            }
+        }
+    }
+
+  if (position_found)
+    {
+      found_positions.push_back(real_location);
+      return true;
+    }
+  else
+    return false;
+}
+
+template <int dim>
+void
+RPTFEMReconstruction<dim>::read_experimental_counts(
+  std::vector<std::vector<double>> &all_experimental_counts)
+{
+  TimerOutput::Scope t(computing_timer, "read_experimental_counts");
+
+  std::ifstream in(
+    rpt_parameters.fem_reconstruction_param.experimental_counts_file);
+  double value;
+
+  if (in)
+    {
+      std::string line;
+
+      while (std::getline(in, line))
+        {
+          // Ignore empty lines
+          if (line == "")
+            continue;
+
+          all_experimental_counts.emplace_back(std::vector<double>());
+
+          // Break down the row into column values
+          std::stringstream split(line);
+
+          while (split >> value)
+            all_experimental_counts.back().push_back(value);
+        }
+    }
+}
+
+
+template <int dim>
+void
+RPTFEMReconstruction<dim>::trajectory()
+{
+  // Tolerance or the extrapolation limit in the reference space for a found
+  // position
+  double tol_reference_location = 0.005;
```

Comment:
Is there a specific reason for 0.005?

Reply:
It was a little arbitrary, but when testing out, it gave decent results. Should I keep it that way or change it to a parameter that would be specified in the .prm file ?

Reply:
Did you try it with different meshes size? I would change it to something relative to max cell size, or something related to the current mesh size that make sense, but I don't know the best way.
Moving it to .prm is a safer solution, you could keep 0.005 as default value :)

Reply:
I am agree with moving it to the .prm file and keep 0.005 as the default! :)

Reply:
Yes, I did try it for different mesh sizes :)
I'll move it and keep the 0.005 as a default !